# Linear Regression

### Part 1 - Data Preprocessing

### Loading the dataset

In [3]:
import pandas as pd
dataset = pd.read_csv('sales data.csv') 
dataset.head(10)

QUANTITYORDERED  PRICEEACH  STATUS  MONTH_ID  PRODUCTLINE  DEALSIZE  \
0               30      95.70       1         2            1         1   
1               34      81.35       1         5            1         1   
2               41      94.74       1         7            1         2   
3               45      83.26       1         8            1         2   
4               49     100.00       1        10            1         2   
5               36      96.66       1        10            1         2   
6               29      86.13       1        11            1         1   
7               48     100.00       1        11            1         2   
8               22      98.57       1        12            1         1   
9               41     100.00       1         1            1         2   

   TERRITORY    SALES  
0          1  2871.00  
1          2  2765.90  
2          2  3884.34  
3          1  3746.70  
4          1  5205.27  
5          1  3479.76  
6          2  2497.77  
7          2  5512.32  
8          1  2168.54  
9          2  4708.44

### Extracting inputs (X) and output (y)

In [4]:
X = dataset.iloc[:,:-1].values
X

array([[ 30.  ,  95.7 ,   1.  , ...,   1.  ,   1.  ,   1.  ],
       [ 34.  ,  81.35,   1.  , ...,   1.  ,   1.  ,   2.  ],
       [ 41.  ,  94.74,   1.  , ...,   1.  ,   2.  ,   2.  ],
       ...,
       [ 43.  , 100.  ,   5.  , ...,   4.  ,   2.  ,   2.  ],
       [ 34.  ,  62.24,   1.  , ...,   4.  ,   1.  ,   2.  ],
       [ 47.  ,  65.52,   6.  , ...,   4.  ,   2.  ,   1.  ]])

In [5]:
y = dataset.iloc[:,-1].values
y

array([2871.  , 2765.9 , 3884.34, ..., 5417.57, 2116.16, 3079.44])

### Creating the Training Set and the Test Set

In [6]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.2,random_state=0)

In [7]:
X_train

array([[ 45.  ,  92.08,   5.  , ...,   4.  ,   2.  ,   2.  ],
       [ 44.  , 100.  ,   1.  , ...,   2.  ,   2.  ,   2.  ],
       [ 47.  , 100.  ,   1.  , ...,   2.  ,   2.  ,   2.  ],
       ...,
       [ 29.  , 100.  ,   1.  , ...,   6.  ,   2.  ,   2.  ],
       [ 21.  ,  96.31,   5.  , ...,   4.  ,   1.  ,   2.  ],
       [ 38.  ,  89.38,   1.  , ...,   4.  ,   2.  ,   1.  ]])

In [8]:
X_test

array([[ 47.  , 100.  ,   1.  , ...,   2.  ,   2.  ,   2.  ],
       [ 30.  ,  87.78,   1.  , ...,   1.  ,   1.  ,   1.  ],
       [ 39.  ,  96.34,   1.  , ...,   1.  ,   2.  ,   2.  ],
       ...,
       [ 31.  ,  88.34,   1.  , ...,   4.  ,   1.  ,   2.  ],
       [ 15.  ,  36.93,   1.  , ...,   3.  ,   1.  ,   4.  ],
       [ 40.  ,  53.75,   1.  , ...,   2.  ,   1.  ,   1.  ]])

In [9]:
y_train

array([4143.6 , 5325.76, 5195.85, ..., 3520.6 , 2022.51, 3396.44])

In [10]:
y_test

array([ 5464.69,  2633.4 ,  3757.26,  3051.71,  5127.04,  2546.8 ,
        5920.4 ,  1549.8 ,  5239.5 ,  1320.8 ,  2914.11,  2938.5 ,
        8602.92,  2743.04,  1621.62,  3235.18,  6482.85,  3675.86,
        1476.6 ,  3251.34,  2389.68,  2166.14,  4983.14,  1742.4 ,
        2731.8 ,  2792.52,  3329.09,  4071.85,  2381.04,  2509.  ,
        3902.41,  5568.64,  1185.3 ,  3524.73,  3965.66,  1329.4 ,
        3656.  ,  1484.2 ,  4950.33,  3293.24,  1550.72,  5372.57,
        3340.48,  3029.04,  5294.14,  1820.01,  2457.18,  1997.5 ,
        4207.84,  2149.74,  2910.4 ,  4236.14,  4099.68,  2817.87,
        3910.53,  2678.4 ,  2921.7 ,  4615.68,  1935.09,  5590.  ,
        4713.6 ,  5938.14,  1500.75,  2183.25,  3989.76,  2018.4 ,
        6913.8 ,  2218.52,  4492.36,  1536.57,  6563.06,  3765.32,
        1685.28,  2326.4 ,  4186.73,  1853.32,  2819.28,  6876.11,
        2337.75,  2840.  ,   974.82,  1968.3 ,  1831.2 ,  1463.  ,
        7307.42,  6231.91,  3828.3 ,  3699.88,  6916.12,  1649

## Part 2 - Building and training the model

### Creating the Linear Regression model

In [11]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()

### Training the Model

In [12]:
model.fit(X_train, y_train)

LinearRegression()

### Inference

In [13]:
y_pred = model.predict(X_test)
y_pred

array([5463.20911414, 2529.01688071, 4909.32478664, 4362.70540184,
       4579.37530858, 2352.40494937, 4662.01556855, 1531.99319907,
       5580.42047173, 1230.01100217, 2799.11596416, 2374.51499109,
       6885.32302804, 2448.42127589, 1260.12207188, 4426.97897275,
       5982.57822966, 4880.05030156, 1179.41720737, 4251.78544534,
       2463.74538807, 1885.80505251, 4860.21032998, 1806.64333143,
       2280.35751764, 2322.54322985, 4408.49863782, 4768.52775327,
       2262.33051738, 2505.07104447, 4904.51101764, 5193.56822779,
        899.0514714 , 4532.12530405, 4926.07864406,  990.31155379,
       4823.77277516, 1485.72072832, 4521.92149437, 4503.72418837,
       1459.60866862, 3984.7715671 , 3946.43162615, 3979.86730395,
       5342.7044182 , 1534.4975776 , 2282.83261489, 1829.58959193,
       4641.77911676, 2055.01857582, 2135.93889445, 5150.62742444,
       4785.14687157, 2591.65663015, 4841.30693969, 2481.29648447,
       2765.56947107, 5270.61613741, 1923.57576274, 5048.37244

#### Predicting a single data point

1. QUANTITYORDERED
2. PRICEEACH	
3. STATUS	
4. MONTH_ID	
5. PRODUCTLINE		
6. DEALSIZE	
7. TERRITORY


##### SALES: 2871

In [21]:
model.predict([[30,	95.7,	1,	2,	1,	1,	1,]])


array([2731.63094338])

##### SALES: 5512.22

In [18]:
model.predict([[48,	100,	1,	11,	1,	2,	2,]])


array([5567.74156481])

## Part 3: Evaluating the Model

### R-Squared value

In [19]:
from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
r2

0.8269092680288708

### Adjusted R-Squared value

In [20]:
k = X_test.shape[1]
n = X_test.shape[0]
adj_r2 = 1-(1-r2)*(n-1)/(n-k-1)
adj_r2

0.8247339805534706